Documentation/design-docs/first-class-structs.md

```diff
@@ -15,6 +15,261 @@ Primary Objectives
+
+There are three phases in the JIT that make changes to the representation of struct nodes and lclVars:
+
+* Importer
+  * Vector types are normalized to the appropriate `TYP_SIMD*` type. Other struct nodes have `TYP_STRUCT`.
+  * Proposed: all struct-valued nodes that are created with a class handle will retain an index into
+    the `ClassLayout` cache.
+  * Proposed: Also normalize to new types `TYP_STRUCT1`, etc.
+
+* Struct promotion
+  * Fields of promoted structs become separate lclVars (scalar promoted) with primitive types.
+    * This is currently an all-or-nothing choice (either all fields are promoted, or none), and is
+      constrained in the number of fields that can be promoted.
+    * Proposed: Support additional cases. See [Improve Struct Promotion](#Improve-Struct-Promotion)
+
+* Global morph
+  * Some promoted structs are forced to stack, and become “dependently promoted”.
+    * Proposed: promoted structs are forced to stack ONLY if address taken.
+      * This includes removing unnecessary pessimizations of block copies. See [Improve and Simplify Block Assignment Morphing](#Block-Assignments).
+
+  * Call args 
+    * If the struct has been promoted
+      * It is morphed to `GT_FIELD_LIST` if it is passed on the stack, or if it is passed
+        in registers that exactly match the types of its fields.
+        * Proposed: This transformation would be made even if it is passed in non-matching
+          registers. The necessary transformations for correct code generation would be
+          made in `Lowering`.
+      * If it is passed by reference, a copy is made, and that copy is passed according to
+        the appropriate rule below
+
+    * If it is passed in a single register, it is morphed into a `GT_LCL_FLD` node of the appropriate
+      type.
+      * This may involve making a copy, if the size cannot be safely loaded.
+      * Proposed: This would remain a `GT_OBJ` and would be appropriately transformed in `Lowering`,
+        e.g. using `GT_BITCAST`.
+
+    * If is passed in multiple registers
+      * A `GT_FIELD_LIST` is constructed that represents the load of each register using `GT_LCL_FLD`.
+      * Proposed: This would also remain `GT_OBJ` (or `GT_FIELD_LIST` if promoted) and would be transformed
+        to a `GT_FIELD_LIST` with the appropriate load, assemble or extraction code as needed.
+
+    * Otherwise, if it is passed by reference or on the stack, it is kept as `GT_OBJ` or `GT_LCL_VAR`
+      * If passed by reference, the value is either forced to the stack or copied.
+
+It is proposed to add the following transformations in `Lowering`:
+* Transform struct values that are passed to or returned from calls by creating one or more of the following:
+  * `GT_FIELD_LIST` of `GT_LCL_FLD` when the struct is non-enregisterable and is passed in multiple
+    registers.
+  * `GT_BITCAST` when a promoted floating point field of a single-field struct is passed in an integer register.
```

Comment:
`BITCAST` nodes could also be inserted during morph, not clear if it's useful. I suppose some could be CSE candidates.

Reply:
In genral, `BITCAST` nodes should anyway move up in phase ordering. I already have a change that produces such node early, to avoid blocking variable enregistration due to float/int reinterpretation and it does show some CQ improvements. The only problem I've yet to solve is decomposition of the double/long case on 32 bit targets.